File: frontend/src/lib/schemas/graph.ts
import { z } from 'zod'

export const dimensionTypeSchema = z.enum([
  'root',
  'concept',
  'requirement',
  'gap',
  'benefit',
  'drawback',
  'feasibility',
  'flaw',
  'alternative',
  'question',
])
export type DimensionType = z.infer<typeof dimensionTypeSchema>

export const nodePositionSchema = z.object({
  x: z.number(),
  y: z.number(),
})

export const analysisNodeSchema = z.object({
  id: z.string(),
  type: dimensionTypeSchema,
  label: z.string(),
  content: z.string(),
  score: z.number().nullable().optional(),
  parent_id: z.string().nullable().optional(),
  position: nodePositionSchema,
  userPositioned: z.boolean().default(false),
})
export type AnalysisNode = z.infer<typeof analysisNodeSchema>

export const analysisEdgeSchema = z.object({
  id: z.string(),
  source: z.string(),
  target: z.string(),
  label: z.string().optional(),
  type: z.string().optional(),
})
export type AnalysisEdge = z.infer<typeof analysisEdgeSchema>

export const analysisGraphSchema = z.object({
  nodes: z.array(analysisNodeSchema),
  edges: z.array(analysisEdgeSchema),
})
export type AnalysisGraph = z.infer<typeof analysisGraphSchema>

// LLM graph actions
export const nodePayloadSchema = z.object({
  id: z.string(),
  type: dimensionTypeSchema,
  label: z.string(),
  content: z.string(),
  score: z.number().nullable().optional(),
  parent_id: z.string().nullable().optional(),
})

export const addNodeActionSchema = z.object({
  action: z.literal('add'),
  payload: nodePayloadSchema,
})

export const updateNodeActionSchema = z.object({
  action: z.literal('update'),
  payload: z.object({
    id: z.string(),
    label: z.string().optional(),
    content: z.string().optional(),
  }),
})

export const deleteNodeActionSchema = z.object({
  action: z.literal('delete'),
  payload: z.object({ id: z.string() }),
})

export const connectActionSchema = z.object({
  action: z.literal('connect'),
  payload: z.object({
    source: z.string(),
    target: z.string(),
    label: z.string().optional(),
    type: z.string().optional(),
  }),
})

export const llmGraphActionSchema = z.discriminatedUnion('action', [
  addNodeActionSchema,
  updateNodeActionSchema,
  deleteNodeActionSchema,
  connectActionSchema,
])
export type LLMGraphAction = z.infer<typeof llmGraphActionSchema>

File: backend/app/schemas/graph.py
from enum import StrEnum
from typing import Annotated, Literal, Union

from pydantic import BaseModel, Field, field_validator


class DimensionType(StrEnum):
    ROOT = "root"
    CONCEPT = "concept"
    REQUIREMENT = "requirement"
    GAP = "gap"
    BENEFIT = "benefit"
    DRAWBACK = "drawback"
    FEASIBILITY = "feasibility"
    FLAW = "flaw"
    ALTERNATIVE = "alternative"
    QUESTION = "question"


class NodePosition(BaseModel):
    x: float
    y: float


class AnalysisNode(BaseModel):
    id: str
    type: DimensionType
    label: str
    content: str
    score: float | None = None
    parent_id: str | None = None
    position: NodePosition = Field(default_factory=lambda: NodePosition(x=0, y=0))
    userPositioned: bool = False


class AnalysisEdge(BaseModel):
    id: str
    source: str
    target: str
    label: str | None = None
    type: str | None = None


class AnalysisGraph(BaseModel):
    nodes: list[AnalysisNode] = Field(default_factory=list)
    edges: list[AnalysisEdge] = Field(default_factory=list)

    @field_validator("nodes")
    @classmethod
    def validate_node_count(cls, v: list[AnalysisNode]) -> list[AnalysisNode]:
        if len(v) > 200:
            raise ValueError("Graph cannot have more than 200 nodes")
        return v

    @field_validator("edges")
    @classmethod
    def validate_edge_count(cls, v: list[AnalysisEdge]) -> list[AnalysisEdge]:
        if len(v) > 400:
            raise ValueError("Graph cannot have more than 400 edges")
        return v


# --- LLM graph actions ---


class NodePayload(BaseModel):
    id: str
    type: DimensionType
    label: str
    content: str
    score: float | None = None
    parent_id: str | None = None


class AddNodeAction(BaseModel):
    action: Literal["add"]
    payload: NodePayload


class UpdateNodePayload(BaseModel):
    id: str
    label: str | None = None
    content: str | None = None


class UpdateNodeAction(BaseModel):
    action: Literal["update"]
    payload: UpdateNodePayload


class DeleteNodePayload(BaseModel):
    id: str


class DeleteNodeAction(BaseModel):
    action: Literal["delete"]
    payload: DeleteNodePayload


class ConnectPayload(BaseModel):
    source: str
    target: str
    label: str | None = None
    type: str | None = None


class ConnectAction(BaseModel):
    action: Literal["connect"]
    payload: ConnectPayload


LLMGraphAction = Annotated[
    Union[AddNodeAction, UpdateNodeAction, DeleteNodeAction, ConnectAction],
    Field(discriminator="action"),
]


class LLMResponse(BaseModel):
    message: str
    graph_actions: list[LLMGraphAction]
